aws/resource_aws_emr_instance_fleet.go

```diff
@@ -0,0 +1,193 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/emr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func resourceAwsEMRInstanceFleet() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsEMRInstanceFleetCreate,
+		Read:   resourceAwsEMRInstanceFleetRead,
+		Update: resourceAwsEMRInstanceFleetUpdate,
+		Delete: resourceAwsEMRInstanceFleetDelete,
+		Schema: map[string]*schema.Schema{
+			"cluster_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"instance_fleet_type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice(emr.InstanceFleetType_Values(), false),
+			},
+			"instance_fleet": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem:     InstanceFleetConfigSchema(),
+			},
+		},
+	}
+}
+
+func resourceAwsEMRInstanceFleetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+
+	addInstanceFleetInput := &emr.AddInstanceFleetInput{
+		ClusterId: aws.String(d.Get("cluster_id").(string)),
+	}
+
+	if l := d.Get("instance_fleet").([]interface{}); len(l) > 0 && l[0] != nil {
+		addInstanceFleetInput.InstanceFleet = readInstanceFleetConfig(
+			l[0].(map[string]interface{}),
+			*aws.String(d.Get("instance_fleet_type").(string)))
+	}
+
+	log.Printf("[DEBUG] Creating EMR instance fleet params: %s", addInstanceFleetInput)
+	resp, err := conn.AddInstanceFleet(addInstanceFleetInput)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Created EMR instance fleet finished: %#v", resp)
+	if resp == nil {
+		return fmt.Errorf("error creating instance fleet: no instance fleet returned")
+	}
+	d.SetId(*resp.InstanceFleetId)
+
+	return nil
+}
+
+func resourceAwsEMRInstanceFleetRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).emrconn
+	instanceFleets, err := fetchAllEMRInstanceFleets(conn, d.Get("cluster_id").(string))
+	if err != nil {
+		log.Printf("[DEBUG] EMR doesn't have any Instance Fleet ")
+		d.SetId("")
+		return nil
+	}
+
+	fleet := findInstanceFleetById(instanceFleets, d.Id())
+	if fleet == nil {
+		log.Printf("[DEBUG] EMR Instance Fleet (%s) not found, removing", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("instance_fleet_type", aws.StringValue(fleet.InstanceFleetType))
+
+	if err := d.Set("instance_fleet", flattenInstanceFleet(fleet)); err != nil {
+		return fmt.Errorf("error setting instance_fleet: %s", err)
+	}
+
+	return nil
+}
+
+func findInstanceFleetById(instanceFleets []*emr.InstanceFleet, fleetId string) *emr.InstanceFleet {
+	for _, fleet := range instanceFleets {
+		if *fleet.Id == fleetId {
```

Comment:
To prevent potential panics:

```suggestion
		if fleet == nil {
			continue
		}

		if aws.StringValue(fleet.Id) == fleetId {
```